File: Stripe/ApplePay/PKPayment+Stripe.m
//
//  PKPayment+Stripe.m
//  Stripe
//
//  Created by Ben Guo on 7/2/15.
//

#import "PKPayment+Stripe.h"

@implementation PKPayment (Stripe)

- (BOOL)isSimulated {
    return [self.token.transactionIdentifier isEqualToString:@"Simulated Identifier"];
}

- (void)setFakeTransactionIdentifier {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wundeclared-selector"

    PKPaymentToken *token = [PKPaymentToken new];
    if ([token respondsToSelector:@selector(setTransactionIdentifier:)]) {
        NSString *uuid = [[NSUUID UUID] UUIDString];
        uuid = [uuid stringByReplacingOccurrencesOfString:@"~" withString:@""
                                                  options:0
                                                    range:NSMakeRange(0, uuid.length)];
        
        // Simulated cards don't have enough info yet. For now, use a fake Visa number
        NSString *number = @"4242424242424242";

        // Without the original PKPaymentRequest, we'll need to use fake data here.
        NSDecimalNumber *amount = [NSDecimalNumber decimalNumberWithString:@"20.00"];
        NSString *cents = [@([[amount decimalNumberByMultiplyingByPowerOf10:2] integerValue]) stringValue];
        NSString *currency = @"USD";
        NSString *identifier = [@[@"ApplePayStubs", number, cents, currency, uuid] componentsJoinedByString:@"~"];
        [token performSelector:@selector(setTransactionIdentifier:) withObject:identifier];
    }
    if ([self respondsToSelector:@selector(setToken:)]) {
        [self performSelector:@selector(setToken:) withObject:token];
    }
#pragma clang diagnostic pop
}

@end

File: Stripe/PKPayment+Stripe.swift
//
//  PKPayment+Stripe.swift
//  Stripe
//
//  Created by Ben Guo on 7/2/15.
//

import PassKit

extension PKPayment {
  /// Returns true if the instance is a payment from the simulator.
  @objc func stp_isSimulated() -> Bool {
    return token.transactionIdentifier == "Simulated Identifier"
  }

  /// Returns a fake transaction identifier with the expected ~-separated format.
  @objc class func stp_testTransactionIdentifier() -> String? {
    var uuid = UUID().uuidString
    uuid = uuid.replacingOccurrences(of: "~", with: "")

    // Simulated cards don't have enough info yet. For now, use a fake Visa number
    let number = "4242424242424242"

    // Without the original PKPaymentRequest, we'll need to use fake data here.
    let amount = NSDecimalNumber(string: "0")
    let cents = NSNumber(value: amount.multiplying(byPowerOf10: 2).intValue).stringValue
    let currency = "USD"
    let identifier = ["ApplePayStubs", number, cents, currency, uuid].joined(separator: "~")
    return identifier
  }
}
